sources/mma/mma.f90

```diff
@@ -0,0 +1,701 @@
+                                (1.001*max(dfdx(i,j),0.0) + &
+                                0.001*max(-dfdx(i,j),0.0) + &
+                                (0.00001/(max(0.00001, &
+                                (this%xmax(j)-this%xmin(j))))))
+                this%qij(i,j) = (this%x(j)-this%low(j))**2 * &
+                                (0.001*max(dfdx(i,j),0.0) + &
+                                1.001*max(-dfdx(i,j),0.0) + &
+                                (0.00001/(max(0.00001, &
+                                (this%xmax(j)-this%xmin(j))))))
+            end do
+        end do
+
+        !computing bi as defined in page 5
+        do i=1, this%m 
+            this%bi(i) = - fval(i)
+            do j=1, this%n 
+                this%bi(i) = this%bi(i) + &
+                            this%pij(i,j)/ (this%upp(j)-this%x(j)) + &
+                            this%qij(i,j)/(this%x(j)-this%low(j))
+            end do        
+        end do
+
+    end subroutine mma_gensub
+
+    subroutine mma_subsolve_dpip(this)
+        ! ----------------------------------------------------- !
+        ! Dual-primal interior point method using Newton's step !
+        ! to solve MMA sub problem.                             !
+        ! A Backtracking Line Search approach is used to compute!
+        ! the step size; starting with the full Newton's step   !
+        ! (delta=1) and deviding by 2 until we have a step size !
+        ! that leads to a feasible point while ensuring a       !
+        ! decrease in the residue.                              !
+        ! ----------------------------------------------------- !
+        class(mma_t), intent(inout) :: this
+
+        integer :: i, j, iter, ggdumiter, itto
+        real(kind=rp) :: epsi, residumax, residunorm, z, zeta, rez, rezeta, &
+            delz, dz, dzeta, steg, dummy_one, zold, zetaold, newresidu
+        real(kind=rp), dimension(this%m) :: y, lambda, s, mu, &
+            rey, relambda, remu, res, dely, dellambda, dummyGDinv, &
+            dy, dlambda, ds, dmu, yold, lambdaold, sold, muold
+        real(kind=rp), dimension(this%n) :: x, xsi, eta, rex, rexsi, reeta, &
+            delx, diagx, dx, dxsi, deta, xold, xsiold, etaold
+        real(kind=rp), dimension(3*this%n+4*this%m+2) :: residu
+        real(kind=rp), dimension(2*this%n+4*this%m+2) :: xx, dxx
+
+        real(kind=rp), dimension(this%m,this%n) :: GG
+        real(kind=rp), dimension(this%m+1) :: bb
+        real(kind=rp), dimension(this%m+1, this%m+1) :: AA
+        ! using DGESV( N, NRHS, A, LDA, IPIV, B, LDB, INFO ) in lapack to solve
+        ! the linear system which needs the following parameters
+        integer :: info
+        integer, dimension(this%m+1) :: ipiv
+
+        !intial value for the parameters in the subsolve based on 
+        !page 15 of "https://people.kth.se/~krille/mmagcmma.pdf"
+        dummy_one = 1
+        epsi = 1
+        x(:) = 0.5*(this%alpha(:)+this%beta(:))
```

Comment:
There is no reason to specify `(:)`.

```suggestion
        x = 0.5_rp * (this%alpha + this%beta)
```

Reply:
True, but in implementation, this approach helped me track the dimensions of different parameters. I also believe it enhances code readability, especially in long equations involving operations between parameters of various dimensions (arrays of different sizes).
However, if this is a style thingy for NEKO, please let me know and I will remove all (:) and (:,:) in the code.